### PR TITLE
Reduce memory requirements for checkpoint conversion 

### PR DESCRIPTION
**Pull request by khatwanimohit**
# Description

This PR introduces a lazy_load configuration option that reduces peak RAM usage by deferring the loading and transformation of weights until the exact moment Orbax needs to save them to disk.

Note: lazy loading feature is temporarily disabled for multimodal models because the key names in the files are different from what are seen when loading a HF model using AutoConfig.  

## Key Changes

- Lazy Loading Mechanism: Implemented LazyHFLoader and LazyTensor classes. Instead of eagerly loading all weights, we now create lightweight proxy objects that only load their specific tensor data from disk when Orbax calls __array__() during the saving phase.
- Orbax Integration: Registered a custom LazyTensorHandler with Orbax to allow it to correctly recognize and process our proxy objects as if they were standard NumPy arrays.
- Config: Added a new boolean flag lazy_load to the MaxText configuration to enable this mode. It defaults to False to preserve existing behavior for smaller models.

If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/458745828

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

x here denotes number of B parameters

## Llama.3.1-70B

old RAM usage: 616 GB (8.8x GB)
new RAM usage: 86.2 GB (1.2x GB)
Logs: https://paste.googleplex.com/6288510168465408#l=571

## LLama3.1-8B

old RAM usage: 51 GB (6.3x GB)
new RAM usage: 31GB (4x GB)
Logs: https://paste.googleplex.com/6128000672333824
forward pass logits test: https://paste.googleplex.com/5595374018494464

## Qwen3-4B

old RAM usage: 37 GB ( 9.2x GB)
new RAM usage: 15.7 GB ( 3.7x GB )
Logs: https://paste.googleplex.com/4933249042350080#l=543
forward pass logit test: https://paste.googleplex.com/5515770054443008

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/utils/ckpt_conversion/to_maxtext.py [hunk 1]
```diff
@@ -25,6 +25,9 @@
                          Defaults to "./mt_output/".
   scan_layers: (bool) Whether the MaxText model was trained with scanned layers.
                This must match the training configuration of the checkpoint.
+  lazy_load: (bool) If True, uses an on-demand loading strategy to minimize RAM
+             usage during conversion. Recommended if, 2 * model_size (GB) >= system RAM
+             Defaults to False.
```
Comment (hengtaoguo) [inline src/MaxText/utils/ckpt_conversion/to_maxtext.py:30]: This is awesome!!

File: src/MaxText/utils/ckpt_conversion/to_maxtext.py [hunk 2]
```diff
@@ -103,8 +134,173 @@ def format_meter(
...
+    return self.size * self.dtype.itemsize
+
+  @property
+  def itemsize(self):
+    return self.dtype.itemsize
+
+  def __array__(self, dtype=None):
+    # This method is called just-in-time by Orbax when saving this specific leaf.
+    try:
+      arr = self._load_fn()
```
Comment (hengtaoguo) [inline src/MaxText/utils/ckpt_conversion/to_maxtext.py:263]: I'm very curious about what's happening under-the-hood.  
- If we are calling one tensor from the checkpoint, what is happening now? Does it only load that specific shard (0001 out of all shards)of the safetensors? 
- And after this tensor is loaded and transformed, do we release the memory for this safetensor shard? 
- What if a following tensor also belongs to the same shard, and does it required to load the same shard all over again? 

Thanks so much Mohit!

File: src/MaxText/configs/base.yml [hunk 3]
```diff
@@ -67,6 +67,8 @@ checkpoint_storage_use_zarr3: True
 # larger models requires higher concurrent GB for I/O
 # default concurrent gb for PytreeCheckpointHandler is 96GB
 checkpoint_storage_concurrent_gb: 96
+# lazyily load tensors for checkpoint conversion
```
[the remark this answers is not in the document]
Comment (shuningjin) [reply src/MaxText/configs/base.yml]: nit (optional): if this is only used in to_maxtext, maybe we can use local argument there like this
https://github.com/AI-Hypercomputer/maxtext/blob/188c4268ffcf41cb2ee7412239e9732b10a1ccf7/tests/forward_pass_logit_checker.py#L475

File: src/MaxText/utils/ckpt_conversion/to_maxtext.py (hunk 2 above)
Comment (shuningjin) [reply src/MaxText/utils/ckpt_conversion/to_maxtext.py:263]: I am very curious as well.

File: src/MaxText/utils/ckpt_conversion/to_maxtext.py [hunk 4]
```diff
@@ -55,7 +68,18 @@
...
+import torch
+from huggingface_hub import hf_hub_download, list_repo_files
 
+# Try importing safetensors, but don't fail immediately if missing (fallback to torch)
+try:
+  from safetensors import safe_open
+
+  HAS_SAFETENSORS = True
+except ImportError:
+  HAS_SAFETENSORS = False
```
Comment (RissyRan) [inline src/MaxText/utils/ckpt_conversion/to_maxtext.py]: Do we still support torch checkpoint?

File: src/MaxText/utils/ckpt_conversion/to_maxtext.py [hunk 5]
```diff
@@ -103,8 +134,173 @@ def format_meter(
...
+    if "model.safetensors.index.json" in files:
+      self.is_safetensors = True
+      index_file = "model.safetensors.index.json"
+    elif "model.safetensors" in files:
+      self.is_safetensors = True
+      # Single file case
+      self.shard_map = {None: "model.safetensors"}
+      return
+    elif "pytorch_model.bin.index.json" in files:
+      index_file = "pytorch_model.bin.index.json"
```
Comment (RissyRan) [inline src/MaxText/utils/ckpt_conversion/to_maxtext.py]: is this support on purpose? I thought we plan to only support safetensors. Otherwise, the mapping could be different for the same model. We had some bad experience in llama4, and would like to get rid of this if possible.

File: src/MaxText/utils/ckpt_conversion/to_maxtext.py (hunk 2 above)
Comment (RissyRan) [reply src/MaxText/utils/ckpt_conversion/to_maxtext.py:263]: +1, @khatwanimohit could you add some explanation in the code or some pointers? I think this help future reader too!

File: src/MaxText/utils/ckpt_conversion/to_maxtext.py (hunk 4 above)
Comment (khatwanimohit) [reply src/MaxText/utils/ckpt_conversion/to_maxtext.py]: removed all the pytorch loading code

File: src/MaxText/configs/base.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (khatwanimohit) [reply src/MaxText/configs/base.yml]: thank you for suggesting this, removed that parameter from base.yml

File: src/MaxText/utils/ckpt_conversion/to_maxtext.py (hunk 5 above)
Comment (khatwanimohit) [reply src/MaxText/utils/ckpt_conversion/to_maxtext.py]: removed all the pytorch loading code

File: src/MaxText/utils/ckpt_conversion/to_maxtext.py (hunk 2 above)
Comment (khatwanimohit) [reply src/MaxText/utils/ckpt_conversion/to_maxtext.py:263]: I have added more comments around this. 
After loading the index file we identify which tensor belongs to which shard, when `get_tensor` is called, it opens the specific shard file using `safetensors.safe_open` which uses `get_tensor()` to load only that specific tensor from the shard. At this point only 1 tensor's worth of RAM is used.

File: src/MaxText/utils/ckpt_conversion/to_maxtext.py [hunk 6]
```diff
@@ -340,16 +557,34 @@ def main(argv: Sequence[str]) -> None:
...
   max_logging.log(f"Conversion complete. Checkpoint saved to {output_directory}")
 
 
 if __name__ == "__main__":
-  app.run(main)
+  jax.config.update("jax_default_prng_impl", "unsafe_rbg")
+  os.environ["TF_CPP_MIN_LOG_LEVEL"] = "0"  # Suppress TensorFlow logging
+
+  parser = argparse.ArgumentParser()
+  parser.add_argument("--lazy_load_tensors", type=bool, default=False, help="Whether to use lazy loading of HF tensors.")
```
Comment (shuningjin) [inline src/MaxText/utils/ckpt_conversion/to_maxtext.py]: Thanks for the update! use `type=str2bool` instead of `type=bool`, and `required=False`
- otherwise `--lazy_load_tensors=False` is still evaluate to true, as "False" is a non-empty string
- from MaxText.inference_utils import str2bool
- ref: https://github.com/AI-Hypercomputer/maxtext/blob/ce7160339f55b470afdaf6a809da77d163629270/src/MaxText/utils/ckpt_scripts/convert_deepseek_family_ckpt.py#L714

**Review by RissyRan — APPROVED** [review]
Thank you!

File: src/MaxText/utils/ckpt_conversion/to_maxtext.py (hunk 6 above)
Comment (khatwanimohit) [reply src/MaxText/utils/ckpt_conversion/to_maxtext.py]: thank you, updated
